File: mcpoke-server/src/processors/type.ts
/**
 * Processing functions for type data
 */
import { pokeApiClient } from '../api/pokeapi.js';
import { cacheManager } from '../cache/cache-manager.js';
import { capitalizeFirstLetter } from '../utils/formatters.js';
import { TypeResponse } from '../api/types.js';

/**
 * Processed type data structure
 */
export interface ProcessedType {
  id: number;
  name: string;
  damage_relations: {
    double_damage_from: string[];
    double_damage_to: string[];
    half_damage_from: string[];
    half_damage_to: string[];
    no_damage_from: string[];
    no_damage_to: string[];
  };
}

/**
 * Process a type response from the API
 * @param typeData The raw type data
 * @returns The processed type data
 */
export function processTypeData(typeData: TypeResponse): ProcessedType {
  return {
    id: typeData.id,
    name: capitalizeFirstLetter(typeData.name),
    damage_relations: {
      double_damage_from: typeData.damage_relations.double_damage_from.map(type => type.name),
      double_damage_to: typeData.damage_relations.double_damage_to.map(type => type.name),
      half_damage_from: typeData.damage_relations.half_damage_from.map(type => type.name),
      half_damage_to: typeData.damage_relations.half_damage_to.map(type => type.name),
      no_damage_from: typeData.damage_relations.no_damage_from.map(type => type.name),
      no_damage_to: typeData.damage_relations.no_damage_to.map(type => type.name)
    }
  };
}

/**
 * Calculate the effectiveness of an attacking type against defending types
 * @param attackingType The attacking type
 * @param defendingTypes The defending types
 * @returns The type effectiveness data
 */
export async function calculateTypeEffectiveness(
  attackingType: string,
  defendingTypes: string[]
): Promise<{
  attacking_type: string;
  defending_types: string[];
  effectiveness: number;
  description: string;
}> {
  // Get attacking type data
  const typeData = await cacheManager.getOrSet<TypeResponse>(
    `type-${attackingType.toLowerCase()}`,
    () => pokeApiClient.getType(attackingType.toLowerCase())
  );
  
  const processedType = processTypeData(typeData);
  
  // Calculate effectiveness for each defending type
  let effectiveness = 1.0;
  
  for (const defendingType of defendingTypes) {
    const lowerDefType = defendingType.toLowerCase();
    
    if (processedType.damage_relations.double_damage_to.includes(lowerDefType)) {
      effectiveness *= 2.0;
    } else if (processedType.damage_relations.half_damage_to.includes(lowerDefType)) {
      effectiveness *= 0.5;
    } else if (processedType.damage_relations.no_damage_to.includes(lowerDefType)) {
      effectiveness = 0;
      break; // No need to continue if there's no effect
    }
  }
  
  // Determine description based on effectiveness
  let description = 'Normally effective';
  
  if (effectiveness === 0) {
    description = 'No effect';
  } else if (effectiveness < 1) {
    description = 'Not very effective';
  } else if (effectiveness > 1) {
    description = 'Super effective';
  }
  
  return {
    attacking_type: capitalizeFirstLetter(attackingType),
    defending_types: defendingTypes.map(type => capitalizeFirstLetter(type)),
    effectiveness,
    description
  };
}

File: src/processors/type.py
"""
Type data processor.

This module processes type data from the PokéAPI.
"""

import logging
from typing import Dict, Any, List

from ..api_client import PokeAPIClient
from ..cache import CacheManager

logger = logging.getLogger('mcpoke-server.processors.type')

class TypeProcessor:
    """
    Processor for type data.
    
    This class processes type data from the PokéAPI.
    """
    
    def __init__(self, api_client: PokeAPIClient, cache: CacheManager):
        """
        Initialize the type processor.
        
        Args:
            api_client: PokéAPI client
            cache: Cache manager
        """
        self.api_client = api_client
        self.cache = cache
    
    def get_type(self, name_or_id: str) -> Dict[str, Any]:
        """
        Get processed information about a type.
        
        Args:
            name_or_id: Name or ID of the type
        
        Returns:
            Processed type information
        """
        # This is a skeleton implementation. We'll just pass for now.
        pass
    
    def get_type_effectiveness(self, attacking_type: str, defending_types: List[str]) -> Dict[str, Any]:
        """
        Calculate type effectiveness.
        
        Args:
            attacking_type: Attacking type
            defending_types: List of defending types
        
        Returns:
            Type effectiveness information
        """
        # This is a skeleton implementation. We'll just pass for now.
        pass
    
    def _process_type_data(self, type_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Process raw type data.
        
        Args:
            type_data: Raw type data from the API
        
        Returns:
            Processed type information
        """
        # This is a skeleton implementation. We'll just pass for now.
        pass
    
    def _calculate_effectiveness(self, attacking_type_data: Dict[str, Any], defending_types_data: List[Dict[str, Any]]) -> float:
        """
        Calculate the effectiveness of an attack against a combination of defending types.
        
        Args:
            attacking_type_data: Raw attacking type data
            defending_types_data: List of raw defending type data
        
        Returns:
            Effectiveness multiplier
        """
        # This is a skeleton implementation. We'll just pass for now.
        pass
    
    def _get_effectiveness_description(self, effectiveness: float) -> str:
        """
        Get a description of the effectiveness.
        
        Args:
            effectiveness: Effectiveness multiplier
        
        Returns:
            Description string
        """
        # This is a skeleton implementation. We'll just pass for now.
        pass
